internal/blocksync/reactor.go

```diff
@@ -553,6 +439,154 @@ FOR_LOOP:
+	if err != nil {
+		// TODO This is bad, are we zombie?
+		panic(fmt.Sprintf("Failed to process committed block (%d:%X): %v", first.Height, first.Hash(), err))
+	}
+	bcR.metrics.recordBlockMetrics(first)
+	*blocksSynced++
+
+	if *blocksSynced%100 == 0 {
+		*lastRate = 0.9**lastRate + 0.1*(100/time.Since(*lastHundred).Seconds())
+		bcR.Logger.Info("Block Sync Rate", "height", bcR.pool.height,
+			"max_peer_height", bcR.pool.MaxPeerHeight(), "blocks/s", *lastRate)
+		*lastHundred = time.Now()
+	}
+
+	return state, err
+}
+
+// switchToConsensus checks if the node is caught up with the rest of the network
+// and switches to consensus reactor if it is. It also logs syncing statistics.
+func (bcR *Reactor) switchToConsensus(state sm.State, initialCommitHasExtensions bool, blocksSynced uint64, stateSynced bool) {
+	height, numPending, lenRequesters := bcR.pool.GetStatus()
+	outbound, inbound, _ := bcR.Switch.NumPeers()
+	bcR.Logger.Debug("Consensus ticker", "numPending", numPending, "total", lenRequesters,
+		"outbound", outbound, "inbound", inbound, "lastHeight", state.LastBlockHeight)
+
+	// The "if" statement below is a bit confusing, so here is a breakdown
+	// of its logic and purpose:
+	//
+	// If we are at genesis (no block in the chain), we don't need VoteExtensions
+	// because the first block's LastCommit is empty anyway.
+	//
+	// If VoteExtensions were disabled for the previous height then we don't need
+	// VoteExtensions.
+	//
+	// If we have sync'd at least one block, then we are guaranteed to have extensions
+	// if we need them by the logic inside loop FOR_LOOP: it requires that the blocks
+	// it fetches have extensions if extensions were enabled during the height.
+	//
+	// If we already had extensions for the initial height (e.g. we are recovering),
+	// then we are guaranteed to have extensions for the last block (if required) even
+	// if we did not blocksync any block.
+	//
+	missingExtension := true
+	if state.LastBlockHeight == 0 ||
+		!state.ConsensusParams.ABCI.VoteExtensionsEnabled(state.LastBlockHeight) ||
+		blocksSynced > 0 ||
+		initialCommitHasExtensions {
+		missingExtension = false
+	}
+
+	// If require extensions, but since we don't have them yet, then we cannot switch to consensus yet.
+	if missingExtension {
+		bcR.Logger.Info(
+			"no extended commit yet",
+			"height", height,
+			"last_block_height", state.LastBlockHeight,
+			"initial_height", state.InitialHeight,
+			"max_peer_height", bcR.pool.MaxPeerHeight(),
+		)
+		return
+	}
+	if bcR.pool.IsCaughtUp() {
+		bcR.Logger.Info("Time to switch to consensus mode!", "height", height)
+		if err := bcR.pool.Stop(); err != nil {
+			bcR.Logger.Error("Error stopping pool", "err", err)
+		}
+		if memR, ok := bcR.Switch.Reactor("MEMPOOL").(mempoolReactor); ok {
+			memR.EnableInOutTxs()
+		}
+		if conR, ok := bcR.Switch.Reactor("CONSENSUS").(consensusReactor); ok {
+			conR.SwitchToConsensus(state, blocksSynced > 0 || stateSynced)
+		}
+	}
```

Comment:
however here, there was a `break FOR_LOOP`, which would stop the loop outside the `select`, however here, the loop would continue to execute, just as in the previous `return`, which is not the current behavior. Am I missing anything here?

Reply:
I'm going to need to merge main here, and then I should be able to answer this in more detail.  

I definitely remember playing with the labels.

Reply:
Ok, so I got myself a bit more oriented, and I don't think that I replaced the "break".  Do you have any ideas on what we might do to solve this?

Also, since this is passing tests, should we look to implement another test?  

(I guess, if the tests are rigorous enough then the answer here is "no worries"?)

Reply:
Thanks, @faddat 
Will review it